lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart

```diff
@@ -718,3 +864,85 @@ class MutatorsStack extends Iterable<Mutator> {
   @override
   Iterator<Mutator> get iterator => _mutators.reversed.iterator;
 }
+
+/// The results of diffing the current composition order with the active
+/// composition order.
+class ViewListDiffResult {
+  /// Views which should be removed from the scene.
+  final List<int> viewsToRemove;
+
+  /// Views to add to the scene.
+  final List<int> viewsToAdd;
+
+  /// If `true`, [viewsToAdd] should be added at the beginning of the scene.
+  /// Otherwise, they should be added at the end of the scene.
+  final bool addToBeginning;
+
+  /// If [addToBeginning] is `true`, then this is the id of the platform view
+  /// to insert [viewsToAdd] before.
+  ///
+  /// `null` if [addToBeginning] is `false`.
+  final int? viewToInsertBefore;
+
+  const ViewListDiffResult(
+      this.viewsToRemove, this.viewsToAdd, this.addToBeginning,
+      {this.viewToInsertBefore});
+}
+
+/// Diff the composition order with the active composition order. It is
+/// common for the composition order and active composition order to differ
+/// only slightly.
+///
+/// Consider a scrolling list of platform views; from frame
+/// to frame the composition order will change in one of two ways, depending
+/// on which direction the list is scrolling. One or more views will be added
+/// to the beginning of the list, and one or more views will be removed from
+/// the end of the list, with the order of the unchanged middle views
+/// remaining the same.
```

Comment:
Checkout [_insertChildDomNodes](https://github.com/flutter/engine/blob/d877a4bbcd0e5b48d2ff18fcba7e010dc52276fe/lib/web_ui/lib/src/engine/html/surface.dart#L1035). With the same amount of code (reusing `longestIncreasingSubsequence`) we should be able to find the optimal solution for all situations, including the scroll case.

Reply:
Added a TODO. That refactor would also necessitate copying more code, for instance to generate the index maps.